elastic-agent-client.proto

```diff
@@ -112,15 +112,190 @@ message ActionResponse {
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Discards the transaction in the store.
+message StoreDiscardRequest {
+    // Token that is used to unique identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+message ArtifactFetchRequest {
+    // Token that is used to unique identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+    // ID of the artifact.
+    string id = 2;
+    // SHA256 of the artifact.
+    string sha256 = 3;
+}
+
+message ArtifactFetchResponse {
+    // Artifact content.
+    bytes content = 1;
+}
+
+message UnitMetric {
+    // ID of the unit.
+    string id = 1;
+
+    // TODO: metrics content
+}
+
+message MetricsReportRequest {
+    // Token that is used to unique identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+
+    // TODO: metrics content
+
+    repeated UnitMetric units = 2;
+}
+
 service ElasticAgent {
     // Called by the client to provide the Elastic Agent the state of the application.
     //
-    // A `StateObserved` must be streamed at least every 30 seconds or it will result in the
-    // application be automatically marked as FAILED, and after 60 seconds it will be force killed and
-    // restarted.
-    rpc Checkin(stream StateObserved) returns (stream StateExpected);
+    // A `CheckinObserved` must be streamed at least every 30 seconds or it will result in the
+    // set of units automatically marked as FAILED, and after 60 seconds it will be force killed the entire
+    // process and restart it.
+    rpc Checkin(stream CheckinObserved) returns (stream CheckinExpected);
 
     // Called by the client on connection to the GRPC allowing the Elastic Agent to stream action
     // requests to the application and the application stream back responses to those requests.
+    //
+    // Request and response is swapped here because the Elastic Agent sends the requests in a stream
+    // to the connected process. The order of response from the process does not matter, it is acceptable
+    // for the response order to be different then the request order.
     rpc Actions(stream ActionResponse) returns (stream ActionRequest);
+
+    // Key-Value state storage is provided for each unit.
+    //
+    // Transactional store is provided to allow multiple key operations to occur before a commit to ensure consistent
+    // state when multiple keys make up the state of an units persistent state.
+    rpc StoreBeginTxn(StoreBeginTxnRequest) returns (StoreBeginTxnResponse);
+    rpc StoreGetKey(StoreGetKeyRequest) returns (StoreGetKeyResponse);
+    rpc StoreSetKey(StoreSetKeyRequest) returns (EmptyResponse);
+    rpc StoreDeleteKey(StoreDeleteKeyRequest) returns (EmptyResponse);
+    rpc StoreCommit(StoreCommitRequest) returns (EmptyResponse);
```

Comment:
Note that if we decide in the future to return something other than an empty message, changing the return message type is not backwards compatible.

The way to guard against this is to define an empty `StoreCommitResponse` message as the return type, so that you can safely add fields to it in the future if you need to.

Reply:
Thats a good idea, will do that for all the other empty responses so in the future if return needs to be adjusted it can.